Separate MeshContexts for received and provided meshes

Ziaad-Khaled opened the pull request:
## Main changes of this PR
### Problem
Original `MeshContext` used a boolean flag to distinguish provided/received meshes, causing bug-prone code with conditionally-valid members.

### Solution

#### Type Hierarchy
- Split `MeshContext` into a `struct` base with common members.
- Created `ProvidedMeshContext` and `ReceivedMeshContext` derived structs.
- Removed all virtual methods and `static_cast` usage.
- **Design Note**: The `partition` member is intentionally placed in derived classes (as `std::shared_ptr<ProvidedPartition>` and `std::shared_ptr<ReceivedPartition>`) to maintain type safety and avoid runtime type checks.

#### Storage
- Separate containers: `_providedMeshContexts` and `_receivedMeshContexts` (using `std::deque` for pointer stability).
- `_usedMeshContexts` is now `std::vector<std::variant<ProvidedMeshContext*, ReceivedMeshContext*>>` to maintain global ordering.

#### Call-Site Approach
- **Ordering-Critical Sections (Deadlock Prevention):** Used `std::variant` loop with `std::holds_alternative` and `std::get` (e.g., `compareBoundingBoxes`, `communicate`). This ensures all participants iterate in the same global order (by mesh name).
- **Non-Critical Sections:** Used separate loops (one for provided, one for received) or helper lambdas to simplify code where ordering is not strictly required.
- **Helpers:** Added `getMeshName` and `getMeshContext` helpers using `std::visit` for cleaner variant access.

## Motivation and additional information
closes #1489


## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [ ] I added a test to cover the proposed changes in our test suite.
* [ ] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I stuck to C++17 features.
* [x] I stuck to CMake version 3.22.1.
* [x] I squashed / am about to squash all commits that should be seen as one.
* [ ] I ran the systemtests by adding the label `trigger-system-tests` (may be skipped if minor change)

## Reviewers' checklist

<!-- Tag people next to each point and add points for specific questions -->

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

<!-- add more questions/tasks if necessary -->

fsimonis commented on src/precice/impl/ParticipantState.cpp:
```diff
@@ -91,17 +101,17 @@ void ParticipantState::receiveMesh(const mesh::PtrMesh
   checkDuplicatedUse(meshName);
   PRECICE_ASSERT(!fromParticipant.empty());
   PRECICE_ASSERT(safetyFactor >= 0);
-  auto context               = new MeshContext();
```
This should ideally not allocate but add this object to the `_receivedMeshContexts`.

As this object contains a shared pointer, you may want to look into move semantics.

If _recievedMeshContexts guarantees persistent element locations in memory (aka no-iterator invalidation), then you can add a pointer to the map.
It is a question of there should be one map per type or one unified map. This may need some thought.

fsimonis commented on src/precice/impl/MeshContext.hpp:
```diff
@@ -41,12 +42,6 @@ struct MeshContext {
   /// need to shift this into the MappingContext
   std::shared_ptr<mesh::BoundingBox> userDefinedAccessRegion;
```
This only exists for received meshes

Ziaad-Khaled commented on src/precice/impl/MeshContext.hpp:
```diff
@@ -41,12 +42,6 @@ struct MeshContext {
-  partition::ReceivedPartition::GeometricFilter geoFilter = partition::ReceivedPartition::GeometricFilter::UNDEFINED;
-
   /// Partition creating the parallel decomposition of the mesh
   partition::PtrPartition partition;
```
We sometimes loop over all `_usedMeshContexts`, for example in `src/precice/impl/ParticipantImpl.cpp:1561`:

```cpp
for (MeshContext *meshContext : _accessor->usedMeshContexts()) {
  meshContext->partition->compareBoundingBoxes();
}
``` 
Since we iterate over the base MeshContext here, we cannot completely remove the base usage. I could replace this with two separate loops (one for providedMeshContexts and one for receivedMeshContexts), but I don't think this is the right solution as it duplicates logic and makes the code more verbose.

Should I keep a virtual accessor (e.g., getPartition()) in the base class that points to the specific partition members in the derived classes? This makes `MeshContext` abstract, which breaks tests like `DataContextTest.cpp` that directly instantiate MeshContext. However, production code never instantiates MeshContext directly, only tests do. What do you think could be a better solution?

fsimonis commented on src/precice/impl/ProvidedMeshContext.cpp:
We don't keep empty translation units. Let's remove it.

fsimonis commented on src/precice/impl/MeshContext.hpp:
```diff
@@ -6,13 +6,14 @@
+/// Base class storing common mesh-related objects and data.
+/// ProvidedMeshContext and ReceivedMeshContext derive from this.
 struct MeshContext {
+public:
```
```suggestion
```

Let's avoid visibility in structs.

fsimonis commented on src/precice/impl/ReceivedMeshContext.hpp:
```diff
@@ -0,0 +1,56 @@
+  /// Filters vertices to those within the local access region
+  std::vector<std::reference_wrapper<const mesh::Vertex>> filterVerticesToLocalAccessRegion(bool requiresBB) const;
+
+private:
```
```suggestion
```

It can stay public.

fsimonis commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -1591,26 +1607,49 @@ void ParticipantImpl::computePartitions()
     }
   }
 
+  // Helper lambdas to process each type
```
The processing order of the blocks should remain unchanged.

fsimonis commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -237,8 +239,11 @@ void ParticipantImpl::configure(
+    _meshLock.add(context.mesh->getName(), false);
+  }
+  for (const auto &context : _accessor->receivedMeshContexts()) {
+    _meshLock.add(context.mesh->getName(), false);
```
On second thought, using the usedMeshContexts here could be cleaner.

fsimonis reviewed (commented):
I had another look. This is pretty difficult to review.

You could reduce a lot of code with extractor functions that return the Mesh and the Partition of a context variant.

Also some functions are now unused. You could remove them.

Ziaad-Khaled commented on src/precice/impl/ReceivedMeshContext.cpp:
```diff
@@ -0,0 +1,46 @@
+
+void ReceivedMeshContext::checkVerticesInsideAccessRegion(precice::span<const double> coordinates, int meshDim, std::string_view functionName) const
+{
+  if (userDefinedAccessRegion) {
```
I applied all your other requests, except this one.

I tried changing this `if` to an assertion as you suggested, but it caused the `ImplicitWithWaveform` test to fail.

The reason is that `checkVerticesInsideAccessRegion` is called unconditionally by `ParticipantImpl`. However, `userDefinedAccessRegion` is optional and can be null (which is the case in this test).

Using an assertion crashes valid simulations that don't use an access region. We need the `if` to safely skip the check when no region is defined.

> I had another look. This is pretty difficult to review.

I agree, this PR turned out to be more difficult than I initially thought as well. I appreciate your patience and detailed review. I hope the final result meets your expectations!

Please let me know if this explanation is accepted and how you would like to proceed (e.g., if I should go ahead and squash the commits or if you need other changes).

Ziaad-Khaled commented:
I have applied the guard clause suggestion. Thanks for the feedback!

I'm looking forward to your next review.

Ziaad-Khaled commented on src/precice/impl/ParticipantImpl.cpp:
```diff
@@ -822,7 +819,7 @@ void ParticipantImpl::setMeshVertices(
   PRECICE_TRACE(meshName, positions.size(), ids.size());
   PRECICE_REQUIRE_MESH_MODIFY(meshName);
-  MeshContext &context = _accessor->usedMeshContext(meshName);
+  MeshContext &context = _accessor->meshContext(meshName);
```
Just to confirm: your code suggestion shows `meshContext()`, I've implemented it using `providedMeshContext()` as your comment described. Let me know if that's correct.

Ziaad-Khaled commented:
I’ve addressed all the points you mentioned.

> This PR clarifies many places. Looking forward to merging this!

I’m glad I got to participate in this. Your review was very helpful and definitely gave me valuable insights.

Happy to iterate further if you spot anything else. Looking forward to merging this as well!
